### PR TITLE
Energy density estimator itself

### PR DESCRIPTION
**Pull request by PDoakORNL**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- New feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Build related changes
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation or build script changes
- Other (please describe):

### Does this introduce a breaking change?

- Yes
- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/Estimators/OperatorEstBase.h [hunk 1]
```diff
@@ -91,8 +91,12 @@ class OperatorEstBase
 
   virtual void startBlock(int steps) = 0;
 
+  const std::vector<QMCT::RealType>& get_data() const { return data_; }
   std::vector<QMCT::RealType>& get_data() { return data_; }
+
   virtual std::size_t getFullDataSize() const { return data_.size(); }
+  virtual void packData(PooledData<Real>& buffer);
+  virtual void unpackData(PooledData<Real>& buffer);
```
Comment (ye-luo) [inline src/Estimators/OperatorEstBase.h:99]: Please document these two functions explicitly buffer->local or local->buffer.

File: src/Estimators/PerParticleHamiltonianLogger.h [hunk 2]
```diff
@@ -55,10 +54,12 @@ class PerParticleHamiltonianLogger : public OperatorEstBase
 
   void write(CrowdLogValues& values, const std::vector<long>& walkers_ids);
 
+  Real sumOverAll();
```
Comment (ye-luo) [inline src/Estimators/PerParticleHamiltonianLogger.h:57]: `sumOverAll() const`?

File: src/Estimators/PerParticleHamiltonianLogger.h [hunk 3]
```diff
@@ -55,10 +54,12 @@ class PerParticleHamiltonianLogger : public OperatorEstBase
...
+
   int get_block() { return block_; }
 private:
   bool crowd_clone = false;
-  PerParticleHamiltonianLogger  * const rank_estimator_;
+  PerParticleHamiltonianLogger* const rank_estimator_;
```
Comment (ye-luo) [inline src/Estimators/PerParticleHamiltonianLogger.h:62]: Can we use `const OptionalRef<PerParticleHamiltonianLogger>`?

**Comment by prckent** [comment]
General comment on functions such as `sumOverAll` and `packData` : this might not yet be in our coding guide but I would much prefer more verbose names to indicate more about _what_ is being summed, packed etc., and as Ye put, in which direction operations are happening. Makes the code much easier to read by requiring less knowledge of the other objects and member data.   Not an essential change but we do need to improve accessibility.

**Comment by PDoakORNL** [comment]
The purpose of sumOverAll is shown in `test_EnergyDensityEstimator.cpp:140.`  I expanded the documentation somewhat.

The pack and unpack data will be used in the next PR.  It seems that either the MPI communicator would have to travel over the rank scope estimators or they would need deliver either the addresses of their data or a copy of their data to the MPI communicator. I decided that for now a single packed copy with lifetime decoupled from the estimators provided the best compromise between decoupling, performance, and memory use.  The documentation for that will be in the next PR since we haven't made to the enclosing scope or level in the call stack where the buffer lives.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**